designs/s3/design-vscode-s3.md

```diff
@@ -0,0 +1,262 @@
+
+### S3 root node
+
+The S3 root node is the "meta node" which lists available buckets in the
+current region.
+
+- Toolkit toolbar:
+  <br><img src="design-vscode-s3-toolbar.png" width="200"/>
+  - (P0) When any S3 node (bucket/folder/object) is selected, Toolkit toolbar
+    enables an `Upload...` button.
+    - If a non-folder object is selected, `Upload...` targets the object's parent folder.
+- S3 root node context-menu:
+  - (P0) `New bucket`
+    - Invokes the `AWS: New bucket` VSCode command which prompts for Bucket name.
+    - Public access is blocked by default. To enable public access, user must
+      use the AWS web console.
+
+### S3 buckets
+
+Buckets are listed as children of the S3 root node.
+
+- Bucket node context-menu:
+  - (P0) `Copy ARN`
+  - (P0) `Copy name`
+  - (P0) `New folder...`
+  - (P0) `Upload...`
+    - Shows VSCode-standard "Open file..." dialog.
+    - Shows a progress dialog using VSCode conventions.
+    - When upload succeeds or fails, shows a toaster summarizing which uploads succeeded and failed.
+    - (P1) Allows multiple files and folders to be uploaded. Creates S3 folders
+      as needed to match the hierarchy of uploaded folders.
+    - (P2) [Auto-detects Content Type ](https://github.com/aws/aws-toolkit-visual-studio/issues/64)
+  - (P0) `Upload to parent folder...`
+    - Works like `Upload...`, except targeting the parent of the current item.
+      This allows the user to avoid needing to navigate to the parent folder.
+  - _Destructive operations (separator):_
+    - (P1) `Delete bucket`
+      - "Super confirm": similar to AWS web console, show a prompt which
+        requires typing the word "Delete".
+
+### S3 objects
+
+S3 objects and folders are listed as children of their respective parent bucket/folder.
+
+- (P0) Clicking an object in the tree selects it, but does not perform any action.
+- (P1) Double-clicking an item downloads it and opens it in readonly mode (see `Edit` action).
+  - If item is 4MB+, prompt user to ask if download should begin.
+- (P2) User input activates standard VSCode local, as-you-type filtering.
+- S3 object node (files and folders) context-menu:
+  - (P0) `Copy ARN`
+  - (P0) `Copy URL`
+  - (P0) `Copy name`
+  - (P0) `Download`
+    - Does _not_ present a chooser.
+    - Downloads immediately to the OS default location (typically `~/Downloads`).
+    - Present download status as a toaster message containing this info:
+      - Path to the download location.
+        - On completion, the path becomes a linkbutton, which onclick opens the file in VSCode.
+      - `Choose location...` linkbutton.
+        - User can change the download location via this link. This changes the
+          default location for future downloads.
+        - Default download location is also configurable as a global Toolkit
+          option. This is not S3-specific, it is used for all `Download`
+          experiences in the Toolkit.
+      - (P2) Loading bar or updating percentage.
```

Comment:
I'd argue that this should be P1 at least, maybe P0-P1 for a "Download Complete" toast with link to the file

Reply:
Updated to P1.